### PR TITLE
oauth: Add QQ connect login

### PR DESCRIPTION
**Pull request by air-adteam**
Add login support for QQ to Hydro.

### CODE REVIEW

**Comment by air-adteam** [comment]
I have read the CLA Document and I hereby sign the CLA

File: packages/login-with-qq/template/home_security.html [hunk 1]
```diff
@@ -0,0 +1,211 @@
+{% extends "layout/home_base.html" %}
```
Comment (undefined-moe) [inline packages/login-with-qq/template/home_security.html]: Template replacement is NOT allowed in official plugin.

File: packages/login-with-qq/locales/en.yaml [hunk 2]
```diff
@@ -0,0 +1,3 @@
+jump: Automatically jump to the homepage to indicate successful binding. Unbound QQ will redirect to the homepage, but cannot log in.
```
Comment (undefined-moe) [inline packages/login-with-qq/locales/en.yaml]: Do NOT use a common word as i18n id.

File: packages/login-with-qq/index.ts [hunk 3]
```diff
@@ -0,0 +1,70 @@
+//Power by air. <air@adteam.cc>
+import {
+    Context, ForbiddenError, Handler, superagent, SystemModel,
+    TokenModel, UserFacingError, ValidationError, OauthModel, OauthCallbackHandler, 
+} from 'hydrooj';
+
+import { Notification } from '@hydrooj/ui-default';
+
+const { PRIV } = global.Hydro.model.builtin; // 内置 Privilege 权限节点
```
Comment (undefined-moe) [inline packages/login-with-qq/index.ts]: also import from 'hydrooj'

File: packages/login-with-qq/index.ts [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+async function callback(this: Handler, { state, code} ) {
+    const [[appid, secret, url], s] = await Promise.all([
+        SystemModel.getMany([
+            'login-qq.id', 'login-qq.secret', 'server.url',
+        ]),
+        TokenModel.get(state, TokenModel.TYPE_OAUTH),
+    ]);
+    const res = await superagent.get(`https://graph.qq.com/oauth2.0/token?grant_type=authorization_code&client_id=${appid}&client_secret=${secret}&code=${code}&redirect_uri=${url}oauth/qq/callback&fmt=json`)
+        .set('accept', 'application/json')
+        .send();
```
Comment (undefined-moe) [inline packages/login-with-qq/index.ts]: Please read the document of superagent.

File: packages/login-with-qq/index.ts [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+
+export function apply(ctx: Context) {
+    ctx.provideModule('oauth', 'qq', {
+        text: 'Login with QQ',
+        callback,
+        get,
+    });
+    ctx.i18n.load('zh', {
+        'Login with QQ': '使用 QQ 登录',
+    });
```
Comment (undefined-moe) [inline packages/login-with-qq/index.ts:71]: Do not mix i18n.load and *.yaml

File: packages/login-with-qq/index.ts [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+        SystemModel.getMany([
+            'login-qq.id', 'login-qq.secret', 'server.url',
+        ]),
+        TokenModel.get(state, TokenModel.TYPE_OAUTH),
+    ]);
+    const res = await superagent.get(`https://graph.qq.com/oauth2.0/token?grant_type=authorization_code&client_id=${appid}&client_secret=${secret}&code=${code}&redirect_uri=${url}oauth/qq/callback&fmt=json`)
+        .set('accept', 'application/json')
+        .send();
+    const fanhui = res.text;
+    const JSobj_token = JSON.parse(fanhui);
```
Comment (undefined-moe) [inline packages/login-with-qq/index.ts]: Use camelCase

**Review by undefined-moe — CHANGES_REQUESTED** [review]

File: packages/login-with-qq/index.ts (hunk 4 above)
Comment (air-adteam) [reply packages/login-with-qq/index.ts]: About QQ connect, this API does NOT support the use of the send method, only supports constructing url. 
Use send as follows：
`const res = await superagent.get('https://graph.qq.com/oauth2.0/token')
        .set('User-Agent', 'ADTeam-OAuth')
        .set('accept', 'application/json')
        .send({
            grant_type: 'authorization_code',
            client_id: appid,
            client_secret: secret,
            code,
            redirect_uri: '${url}oauth/qq/callback',
            fmt: 'json',
        });
`

When logging in, the API returns:
`callback( {"error":100004,"error_description":"param grant_type is wrong or lost "} );`

Comment (undefined-moe) [reply packages/login-with-qq/index.ts]: Read the documention **CAREFULLY**.

File: packages/login-with-qq/template/home_security.html (hunk 1 above)
Comment (air-adteam) [reply packages/login-with-qq/template/home_security.html]: If I need to add elements at the end of the page, how do I do it.

I'm sorry, I haven't done the front-end before. I'm a bit unfamiliar with the things here. Please give me more guidance.

File: packages/login-with-qq/README.md [hunk 7]
```diff
@@ -0,0 +1,17 @@
+# Login-With-QQ
+API采用QQ互联，需在[QQ互联](https://connect.qq.com/)申请API后，在控制面板内填入appid和appkey，方可使用该插件。
+
+该插件由[air.](https://github.com/air-adteam)制作。
+
+因作者技术能力问题，无法在安全设置页面上添加绑定按钮。请自行参考[hydro前端修改文档](https://hydro.js.org/docs/system/frontend-modify/)，并在最后加入以下代码添加：
```
Comment (pandadtdyy) [inline packages/login-with-qq/README.md]: You can support all oauth binding for Hydro in this pr.

**Review by pandadtdyy — CHANGES_REQUESTED** [review]

File: packages/hydrooj/src/handler/user.ts [hunk 8]
```diff
@@ -471,6 +471,11 @@ class OauthCallbackHandler extends Handler {
             this.session.scope = PERM.PERM_ALL.toString();
             this.response.redirect = '/';
         } else {
+            if (this.user.hasPriv(PRIV.PRIV_USER_PROFILE)) {
```
Comment (undefined-moe) [inline packages/hydrooj/src/handler/user.ts:482]: This logic might lead to unwanted behavior.
Binding an oauth account (with different email and unknown openId) is a dangerous operation, this should be a sudo-required operation and user should be notified (via email, etc).
And be sure to add a page allowing user to manage (or delete) any previously binded accounts.

**Review by undefined-moe — CHANGES_REQUESTED** [review]

File: packages/hydrooj/src/handler/user.ts (hunk 8 above)
Comment (air-adteam) [reply packages/hydrooj/src/handler/user.ts:482]: I have discussed the logic related to OAuth with panda, but when deleting OpenID, it was not possible to distinguish between certain types of OAuth (e.g. Github and Google). Except for single user one click unbinding, a single unbinding operation cannot be completed on the existing database

**Pull request closed without merging**